### PR TITLE
Ghidra decompiler

### PR DESCRIPTION
**Pull request by nimashoghi**

### CODE REVIEW

**Comment by mattgodbolt** [comment]
Thanks for this! Will take a look in a while: sorry it's already been 5 days. (I'm on vacation for another week or so, but maybe someone else can take a look too)

File: lib/tooling/_all.js [hunk 1]
```diff
@@ -26,10 +26,11 @@ export { ClangFormatTool } from './clang-format-tool';
...
 export { CompilerDropinTool } from './compiler-dropin-tool';
+export { GhidraTool } from './ghidra';
 export { LLVMMcaTool } from './llvm-mca-tool';
 export { OSACATool } from './osaca-tool';
 export { PaholeTool } from './pahole-tool';
 export { PvsStudioTool } from './pvs-studio-tool';
 export { ReadElfTool } from './readelf-tool';
 export { StringsTool } from './strings-tool';
-export { x86to6502Tool } from './x86to6502-tool';
+export { x86to6502Tool } from './x86to6502-tool';
```
Comment (RubenRBS) [inline lib/tooling/_all.js:36]:
```suggestion
export { x86to6502Tool } from './x86to6502-tool';

```

File: lib/tooling/ghidra.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2020, Compiler Explorer Authors
```
Comment (RubenRBS) [inline lib/tooling/ghidra.js:1]:
```suggestion
// Copyright (c) 2021, Compiler Explorer Authors
```

File: lib/tooling/ghidra.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+// import {spawn} from "child_process"
```
Comment (RubenRBS) [inline lib/tooling/ghidra.js:25]:
```suggestion
```

File: lib/tooling/ghidra.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+//         const child = spawn("c++filt", ["--no-params"])
+//         child.on("error", (error) => reject(error))
+
+//         const bufs = []
+//         child.stdout.on("data", (chunk) => bufs.push(chunk))
+//         child.stdout.on("end", () => resolve(Buffer.concat(bufs)))
+
+//         child.stdin.write(buffer)
+//         child.stdin.end()
+//     })
```
Comment (RubenRBS) [inline lib/tooling/ghidra.js:55]:
```suggestion
```

File: lib/tooling/ghidra.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+        // const exeDir = path.dirname(this.tool.exe)
+        // const output = await _demange(
+        //     result.stdout.map((line) => line.text).join("\n"),
+        // )
+        // result.stdout = utils.parseOutput(
+        //     output.toString("utf-8"),
+        //     compilationInfo.executableFilename,
+        //     exeDir,
+        // )
```
Comment (RubenRBS) [inline lib/tooling/ghidra.js:163]:
```suggestion
```

**Review by RubenRBS — CHANGES_REQUESTED** [review]
Sorry about the radio silence

**Comment by mattgodbolt** [comment]
This looks like a great start, but as @RubenRBS has pointed out there's a few minor things to look at. From the amount of commented-out code it seems incomplete? Is this the case? Is there anything we need to know about to get this merged in, known limitations or similar?

**Comment by nimashoghi** [comment]
First of all, thanks for the review @RubenRBS! I appreciate the comments.

@mattgodbolt, yes, you are correct that this implementation is incomplete and needs to be adjusted.

My primary limitation in implementing this is my lack of experience with the compiler explorer codebase, but with the implementation I have committed, the limitations are:

1. Ghidra's decompiler must be installed beforehands.
2. Ghidra's decompiler needs SLEIGH (instruction set description) files that need to be downloaded, and the SLEIGHHOME environment variable must point to this installation path.
3. The decompiler tooling only works on binary outputs.
4. Compiler explorer's built-in name demangling must be disabled. This means that Ghidra's output will also have demangled names.


My remaining tasks on this were:

1. Figuring out how Compiler Explorer installs these different applications and making sure Ghidra's decompiler is installed properly. This would also involve making sure that Ghidra has the right SLEIGH (instruction set descriptions) installed for the instruction sets that it would be used on. This is due to limitations # 1 and # 2 above.
2. (Less important) Using c++filt to demangle Ghidra's output symbols for functions. This is due to limitation # 4 described above. The commented code was for this demangler. I wanted to put behind some sort of a flag/setting, but I did not have the time to figure out how this is done with the Compiler Explorer's tooling API.

Finally, I have been a bit busy the past two weeks, but I will review these tasks, as well as @RubenRBS' comments and will push my changes accordingly. Please let me know if you have any comments or suggestions.

**Comment by mattgodbolt** [comment]
Thanks @nimashoghi ! That's great. We can definitely help with 1. and 2. - there is a separate repo for installation and set up of apps etc. 3. is a bit of an issue but we have a number of tools that mandate "binary only", so it wouldn't be the first. 4. is something we can work on.

Thank you! We're here to help when you have time.

**Comment by nimashoghi** [comment]
Ok, I have updated my code, which takes care of limitation #4.

I have also created [a PR in the misc-builder repository](https://github.com/compiler-explorer/misc-builder/pull/11) which takes care of limitations #1 and #2.

Feel free to review everything and provide any feedback that you have.

**Comment by mattgodbolt** [comment]
We seem to have stalled on this. I would like to get support but with https://dogbolt.org/ around I wonder if the need has reduced? I mean, taking source and compiling then decompiling is still valuable, which dogbolt doesn't do... so ... maybe still worth looking in to.

**Comment by mattgodbolt** [comment]
Ok - I would love to have this but I'm going to close this. I'd be delighted to reopen. Ghidfra would be great, but this PR is a long way behind main these days, and we're moving more and more to typescript, which will make things worse.

**Pull request closed without merging**